File: tools/connectome/validate_connectome.py
"""
Scientific Validation Suite for MaleCNS Connectome Runtime Graph.
Validates provenance, real body IDs, CSR integrity, synaptic weight ranges,
and absence of synthetic neurons.
"""
import json
import sys
from pathlib import Path

import numpy as np

def validate_connectome_dir(connectome_dir: Path) -> bool:
    print(f"\n=======================================================")
    print(f"   MaleCNS CONNECTOME VALIDATION: {connectome_dir.name}")
    print(f"=======================================================")

    manifest_file = connectome_dir / "manifest.json"
    neurons_file = connectome_dir / "neurons.json"
    indptr_file = connectome_dir / "indptr.bin"
    indices_file = connectome_dir / "indices.bin"
    weights_file = connectome_dir / "weights.bin"
    signs_file = connectome_dir / "signs.bin"

    for f in [manifest_file, neurons_file, indptr_file, indices_file, weights_file, signs_file]:
        if not f.exists():
            print(f"[FAIL] Missing required asset: {f.name}")
            return False

    with open(manifest_file, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    with open(neurons_file, "r", encoding="utf-8") as f:
        neurons = json.load(f)

    indptr = np.fromfile(indptr_file, dtype=np.uint32)
    indices = np.fromfile(indices_file, dtype=np.uint32)
    weights = np.fromfile(weights_file, dtype=np.float32)
    signs = np.fromfile(signs_file, dtype=np.int8)

    errors = []

    # 1. Provenance check
    prov = manifest.get("provenance")
    if prov != "malecns-real":
        errors.append(f"Invalid provenance: expected 'malecns-real', got '{prov}'")

    dataset = manifest.get("dataset")
    if dataset != "MaleCNS":
        errors.append(f"Invalid dataset name: expected 'MaleCNS', got '{dataset}'")

    # 2. Dimensions check
    n_neurons = len(neurons)
    n_edges = len(indices)

    if manifest.get("neuronCount") != n_neurons:
        errors.append(f"Manifest neuronCount ({manifest.get('neuronCount')}) != neurons.json length ({n_neurons})")
    if manifest.get("synapseCount") != n_edges:
        errors.append(f"Manifest synapseCount ({manifest.get('synapseCount')}) != indices.bin length ({n_edges})")

    if len(indptr) != n_neurons + 1:
        errors.append(f"indptr length ({len(indptr)}) != neuronCount + 1 ({n_neurons + 1})")
    if len(weights) != n_edges:
        errors.append(f"weights length ({len(weights)}) != indices length ({n_edges})")
    if len(signs) != n_edges:
        errors.append(f"signs length ({len(signs)}) != indices length ({n_edges})")

    # 3. Real Body ID and Metadata Integrity Check
    body_ids = set()
    indices_set = set()
    synthetic_count = 0
    visual_count = 0
    cx_count = 0
    descending_count = 0
    vnc_count = 0

    for i, neuron in enumerate(neurons):
        b_id = str(neuron.get("bodyId") or "")
        idx = neuron.get("index")

        # Must be genuine numeric MaleCNS ID
        if not b_id.isdigit():
            synthetic_count += 1
            if synthetic_count <= 5:
                errors.append(f"Neuron at index {i} has non-numeric synthetic body ID: '{b_id}'")

        if b_id in body_ids:
            errors.append(f"Duplicate bodyId found: '{b_id}'")
        body_ids.add(b_id)

        if idx != i:
            errors.append(f"Neuron index mismatch: expected {i}, got {idx}")
        indices_set.add(idx)

        # Check for fake badminton-specific biological neurons
        t = str(neuron.get("type") or "")
        name = str(neuron.get("name") or "")
        if any(f in t.lower() or f in name.lower() for f in ["racket", "badminton", "smashneuron", "mn_strike"]):
            errors.append(f"Illegal synthetic badminton neuron in biological graph: '{name}' / '{t}'")

        # Count populations
        region = neuron.get("region")
        if region == "OpticLobe":
            visual_count += 1
        elif region == "CentralComplex":
            cx_count += 1
        elif region == "Descending":
            descending_count += 1
        elif region == "VNC":
            vnc_count += 1

    # 4. CSR Structure Integrity Check
    if indptr[0] != 0:
        errors.append(f"indptr[0] must be 0, got {indptr[0]}")
    if indptr[-1] != n_edges:
        errors.append(f"indptr[-1] ({indptr[-1]}) must equal n_edges ({n_edges})")

    for i in range(n_neurons):
        start = indptr[i]
        end = indptr[i + 1]
        if start > end:
            errors.append(f"indptr monotonicity violation at index {i}: {start} > {end}")
            break
        if end > n_edges:
            errors.append(f"indptr pointer out of bounds at index {i}: {end} > {n_edges}")
            break

    # 5. Weight & Sign Range Check
    if np.any(np.isnan(weights)) or np.any(np.isinf(weights)):
        errors.append("weights array contains NaN or Infinity")
    if np.any(weights <= 0):
        errors.append("weights array contains non-positive conductance values")

    if np.any(indices >= n_neurons):
        errors.append(f"indices array contains out-of-bounds target index (max {indices.max()} >= {n_neurons})")

    valid_signs = {-1, 0, 1}
    unique_signs = set(np.unique(signs))
    if not unique_signs.issubset(valid_signs):
        errors.append(f"signs array contains invalid signs: {unique_signs - valid_signs}")

    # Output Summary Table
    print(f"Dataset          {manifest.get('dataset')} {manifest.get('datasetVersion')}")
    print(f"Provenance       {prov.upper()}")
    print(f"Neurons          {n_neurons:,}")
    print(f"Edges            {n_edges:,}")
    print()
    print(f"Real body IDs    {len(body_ids):,} / {n_neurons:,}")
    print(f"Synthetic IDs    {synthetic_count}")
    print()
    print(f"Visual neurons   {visual_count:,}")
    print(f"Central Complex  {cx_count:,}")
    print(f"Descending       {descending_count:,}")
    print(f"VNC motor        {vnc_count:,}")
    print()
    print(f"CSR integrity    {'PASS' if len(errors) == 0 else 'FAIL'}")
    print(f"Metadata         {'PASS' if synthetic_count == 0 else 'FAIL'}")
    print(f"Provenance       {'PASS' if prov == 'malecns-real' else 'FAIL'}")
    print("-------------------------------------------------------")

    if errors:
        print(f"[FAIL] Connectome validation failed with {len(errors)} errors:")
        for err in errors[:10]:
            print(f"  - {err}")
        return False

    print("CONNECTOME VALID: Real MaleCNS v1.0 biological graph verified.")
    return True

if __name__ == "__main__":
    repo_root = Path(__file__).resolve().parents[2]
    dirs_to_check = [
        repo_root / "data" / "connectome",
        repo_root / "public" / "data" / "connectome",
    ]
    all_ok = True
    for d in dirs_to_check:
        if not validate_connectome_dir(d):
            all_ok = False

    sys.exit(0 if all_ok else 1)

File: scripts/validate-connectome.mjs
import { readFileSync, existsSync, createReadStream } from "node:fs";
import { resolve, join } from "node:path";
import { spawnSync } from "node:child_process";
import { createHash } from "node:crypto";

function tryPythonValidation() {
  const pyCandidates = [
    resolve(process.cwd(), ".venv", "Scripts", "python.exe"),
    resolve(process.cwd(), ".venv", "bin", "python"),
    "python",
    "python3",
  ];

  for (const py of pyCandidates) {
    try {
      const res = spawnSync(py, ["tools/connectome/validate_connectome.py"], {
        stdio: "inherit",
        cwd: process.cwd(),
      });
      if (res.status === 0) {
        return true;
      }
    } catch {
      // Continue to next candidate
    }
  }
  return false;
}

function computeFileHash(filePath) {
  const fileBuffer = readFileSync(filePath);
  return createHash("sha256").update(fileBuffer).digest("hex");
}

function validateDir(dirPath) {
  const dir = resolve(process.cwd(), dirPath);
  console.log(`\n=======================================================`);
  console.log(`   MaleCNS CONNECTOME VALIDATION: ${dirPath}`);
  console.log(`=======================================================`);

  const manifestFile = join(dir, "manifest.json");
  const neuronsFile = join(dir, "neurons.json");
  const indptrFile = join(dir, "indptr.bin");
  const indicesFile = join(dir, "indices.bin");
  const weightsFile = join(dir, "weights.bin");
  const bioWeightsFile = join(dir, "biologicalWeights.bin");
  const signsFile = join(dir, "signs.bin");

  for (const f of [
    manifestFile,
    neuronsFile,
    indptrFile,
    indicesFile,
    weightsFile,
    bioWeightsFile,
    signsFile,
  ]) {
    if (!existsSync(f)) {
      console.error(`[FAIL] Missing required asset: ${f}`);
      return false;
    }
  }

  const manifest = JSON.parse(readFileSync(manifestFile, "utf8"));
  const neurons = JSON.parse(readFileSync(neuronsFile, "utf8"));

  const toAB = (b) => new Uint8Array(b).buffer;
  const indptr = new Uint32Array(toAB(readFileSync(indptrFile)));
  const indices = new Uint32Array(toAB(readFileSync(indicesFile)));
  const weights = new Float32Array(toAB(readFileSync(weightsFile)));
  const biologicalWeights = new Uint32Array(toAB(readFileSync(bioWeightsFile)));
  const signs = new Int8Array(toAB(readFileSync(signsFile)));

  const errors = [];

  if (manifest.provenance !== "malecns-real") {
    errors.push(
      `Invalid provenance: expected 'malecns-real', got '${manifest.provenance}'`,
    );
  }
  if (manifest.dataset !== "MaleCNS") {
    errors.push(
      `Invalid dataset: expected 'MaleCNS', got '${manifest.dataset}'`,
    );
  }

  const nNeurons = neurons.length;
  const nEdges = indices.length;
  const bioSynapsesTotal = biologicalWeights.reduce((acc, w) => acc + w, 0);

  if (manifest.neuronCount !== nNeurons) {
    errors.push(
      `Manifest neuronCount (${manifest.neuronCount}) != neurons.json length (${nNeurons})`,
    );
  }
  const edgeCount = manifest.edgeCount ?? manifest.synapseCount;
  if (edgeCount !== nEdges) {
    errors.push(
      `Manifest edgeCount (${edgeCount}) != indices length (${nEdges})`,
    );
  }
  if (
    manifest.biologicalSynapseTotal !== undefined &&
    manifest.biologicalSynapseTotal !== bioSynapsesTotal
  ) {
    errors.push(
      `Manifest biologicalSynapseTotal (${manifest.biologicalSynapseTotal}) != sum(biologicalWeights) (${bioSynapsesTotal})`,
    );
  }
  if (indptr.length !== nNeurons + 1) {
    errors.push(
      `indptr length (${indptr.length}) != neuronCount + 1 (${nNeurons + 1})`,
    );
  }
  if (weights.length !== nEdges) {
    errors.push(
      `weights length (${weights.length}) != indices length (${nEdges})`,
    );
  }
  if (biologicalWeights.length !== nEdges) {
    errors.push(
      `biologicalWeights length (${biologicalWeights.length}) != indices length (${nEdges})`,
    );
  }
  if (signs.length !== nEdges) {
    errors.push(`signs length (${signs.length}) != indices length (${nEdges})`);
  }

  const bodyIds = new Set();
  let syntheticCount = 0;
  let visualCount = 0;
  let cxCount = 0;
  let descendingCount = 0;
  let vncCount = 0;
  let protocerebrumCount = 0;

  for (let i = 0; i < nNeurons; i++) {
    const neuron = neurons[i];
    const bId = String(neuron.bodyId || "");
    if (!/^\d+$/.test(bId)) {
      syntheticCount++;
      if (syntheticCount <= 5) {
        errors.push(
          `Neuron at index ${i} has non-numeric synthetic body ID: '${bId}'`,
        );
      }
    }
    if (bodyIds.has(bId)) {
      errors.push(`Duplicate bodyId found: '${bId}'`);
    }
    bodyIds.add(bId);

    if (neuron.index !== i) {
      errors.push(`Neuron index mismatch: expected ${i}, got ${neuron.index}`);
    }

    const t = String(neuron.type || "").toLowerCase();
    const name = String(neuron.name || "").toLowerCase();
    if (
      t.includes("racket") ||
      t.includes("badminton") ||
      t.includes("smashneuron") ||
      name.includes("racket")
    ) {
      errors.push(
        `Illegal synthetic badminton neuron in biological graph: '${neuron.name}'`,
      );
    }

    if (neuron.region === "OpticLobe") visualCount++;
    else if (neuron.region === "CentralComplex") cxCount++;
    else if (neuron.region === "Descending") descendingCount++;
    else if (neuron.region === "VNC") vncCount++;
    else if (neuron.region === "Protocerebrum") protocerebrumCount++;
  }

  if (indptr[0] !== 0) errors.push(`indptr[0] must be 0, got ${indptr[0]}`);
  if (indptr[nNeurons] !== nEdges)
    errors.push(
      `indptr[${nNeurons}] (${indptr[nNeurons]}) must equal nEdges (${nEdges})`,
    );

  for (let i = 0; i < nNeurons; i++) {
    if (indptr[i] > indptr[i + 1]) {
      errors.push(`indptr monotonicity violation at ${i}`);
      break;
    }
  }

  for (let e = 0; e < nEdges; e++) {
    if (indices[e] >= nNeurons) {
      errors.push(`indices[${e}] (${indices[e]}) >= neuronCount (${nNeurons})`);
      break;
    }
    if (isNaN(weights[e]) || !isFinite(weights[e]) || weights[e] <= 0) {
      errors.push(`Invalid weight at edge ${e}: ${weights[e]}`);
      break;
    }
    if (biologicalWeights[e] <= 0) {
      errors.push(
        `Invalid biological weight at edge ${e}: ${biologicalWeights[e]}`,
      );
      break;
    }
    if (signs[e] !== -1 && signs[e] !== 0 && signs[e] !== 1) {
      errors.push(`Invalid sign at edge ${e}: ${signs[e]}`);
      break;
    }
  }

  console.log(
    `Dataset          ${manifest.dataset} ${manifest.datasetVersion}`,
  );
  console.log(`Provenance       ${String(manifest.provenance).toUpperCase()}`);
  console.log(`Real Neurons     ${nNeurons.toLocaleString()}`);
  console.log(`Biological Edges ${nEdges.toLocaleString()}`);
  console.log(
    `Total Synapses   ${bioSynapsesTotal.toLocaleString()} biological synaptic contacts`,
  );
  console.log();
  console.log(
    `Real body IDs    ${bodyIds.size.toLocaleString()} / ${nNeurons.toLocaleString()}`,
  );
  console.log(`Synthetic IDs    ${syntheticCount}`);
  console.log();
  console.log(`Visual neurons   ${visualCount.toLocaleString()}`);
  console.log(`Central Complex  ${cxCount.toLocaleString()}`);
  console.log(`Protocerebrum    ${protocerebrumCount.toLocaleString()}`);
  console.log(`Descending       ${descendingCount.toLocaleString()}`);
  console.log(`VNC motor        ${vncCount.toLocaleString()}`);
  console.log();
  console.log(`CSR integrity    ${errors.length === 0 ? "PASS" : "FAIL"}`);
  console.log(`Metadata         ${syntheticCount === 0 ? "PASS" : "FAIL"}`);
  console.log(
    `Provenance       ${manifest.provenance === "malecns-real" ? "PASS" : "FAIL"}`,
  );
  console.log(`-------------------------------------------------------`);

  if (errors.length > 0) {
    console.error(
      `[FAIL] Connectome validation failed with ${errors.length} errors:`,
    );
    for (const err of errors.slice(0, 10)) {
      console.error(`  - ${err}`);
    }
    return false;
  }

  console.log("CONNECTOME VALID: Real MaleCNS v1.0 biological graph verified.");
  return true;
}

if (!tryPythonValidation()) {
  const dirs = ["data/connectome", "public/data/connectome"];
  let allOk = true;
  for (const d of dirs) {
    if (!validateDir(d)) allOk = false;
  }
  if (!allOk) {
    process.exit(1);
  }
}
